pandas/tseries/tools.py

```diff
@@ -510,6 +510,9 @@ def coerce(values):
         # we allow coercion to if errors allows
         return to_numeric(values, errors=errors)
 
+    # prevent overflow in case of int8 or int16
+    arg = arg.astype('int64', copy=False)
```

Comment:
This should cause another problem. On current master:

```
df = pd.DataFrame({'year': [2000, 2001], 'month': [1.5, 1], 'day': [1, 1]})
pd.to_datetime(df)
# ValueError: cannot assemble the datetimes: 'float' object is unsliceable

pd.to_datetime(df.astype(np.int64))
# 0   2000-01-01
# 1   2001-01-01
# dtype: datetime64[ns]
```

Better to raise if input contains non-integer dtype.